### DISCUSSION TITLE
Error when compiling LDT after success for LIS

### DISCUSSION THREAD

**Discussion by tais958** [NASA-LIS/LISF#1056] [Q&A] [answered]
Hi, I was able to successfully compile LIS. Since most of required libraries are identical to what LIS requires, I thought it would go smoothly. However, an error still showed in the process as below:

ifort -c -O2 -u -traceback -fpe0  -nomixed_str_len_arg -names lowercase -convert big_endian -assume byterecl -DIFC -I/qfs/projects/ops/rh6/esmf/7.1/mod/modO/Linux.intel.64.intelmpi.default -DUSE_INCLUDE_MPI -I/pic/projects/sooty2/tais958/lib/eccodes-2.12.5_intel-16.0.2/include/ -I/share/apps/netcdf/4.3.2/intel/16.0.2/include/  -I./ -I../lib/xmrg/ -I../params/VIC/ -I../USAFSI/ ../USAFSI/USAFSI_amsr2Mod.F90
../USAFSI/USAFSI_amsr2Mod.F90(241): error #6683: A kind type parameter must be a compile-time constant.   [HSIZE_T]
      integer(hsize_t), dimension(2)           :: dims_lat, maxdims_lat
--------------^
../USAFSI/USAFSI_amsr2Mod.F90(242): error #6683: A kind type parameter must be a compile-time constant.   [HSIZE_T]
      integer(hsize_t), dimension(2)           :: dims_lon, maxdims_lon
--------------^
../USAFSI/USAFSI_amsr2Mod.F90(243): error #6683: A kind type parameter must be a compile-time constant.   [HSIZE_T]
      integer(hsize_t), dimension(3)           :: dims_tb, maxdims_tb
--------------^
../USAFSI/USAFSI_amsr2Mod.F90(244): error #6683: A kind type parameter must be a compile-time constant.   [HSIZE_T]
      integer(hsize_t), dimension(1)           :: dims_time
--------------^
../USAFSI/USAFSI_amsr2Mod.F90(245): error #6683: A kind type parameter must be a compile-time constant.   [HSIZE_T]
      integer(hsize_t), dimension(2)           :: maxdims_time
--------------^
../USAFSI/USAFSI_amsr2Mod.F90(246): error #6683: A kind type parameter must be a compile-time constant.   [HID_T]
      integer(hid_t)                           :: file_id, gr_id
--------------^
../USAFSI/USAFSI_amsr2Mod.F90(247): error #6683: A kind type parameter must be a compile-time constant.   [HID_T]
      integer(hid_t)                           :: dspace_lat_id, dspace_lon_id
--------------^
../USAFSI/USAFSI_amsr2Mod.F90(248): error #6683: A kind type parameter must be a compile-time constant.   [HID_T]
      integer(hid_t)                           :: dspace_tb_id
--------------^
../USAFSI/USAFSI_amsr2Mod.F90(249): error #6683: A kind type parameter must be a compile-time constant.   [HID_T]
      integer(hid_t)                           :: dspace_time_id
--------------^
../USAFSI/USAFSI_amsr2Mod.F90(250): error #6683: A kind type parameter must be a compile-time constant.   [HID_T]
      integer(hid_t)                           :: lat_id, lon_id
--------------^
../USAFSI/USAFSI_amsr2Mod.F90(251): error #6683: A kind type parameter must be a compile-time constant.   [HID_T]
      integer(hid_t)                           :: tb10h_id, tb10v_id, tb19h_id, tb19v_id, &
--------------^
../USAFSI/USAFSI_amsr2Mod.F90(254): error #6683: A kind type parameter must be a compile-time constant.   [HID_T]
      integer(hid_t)                           :: time_id
--------------^
../USAFSI/USAFSI_amsr2Mod.F90(270): error #6404: This name does not have a type, and must have an explicit type.   [H5F_ACC_RDONLY_F]
      call h5fopen_f(trim(filename),H5F_ACC_RDONLY_F, file_id, status)
------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(376): error #6404: This name does not have a type, and must have an explicit type.   [H5T_IEEE_F32LE]
      call h5dread_f(lat_id, H5T_IEEE_F32LE, lat_raw, dims_lat, status)
-----------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(382): error #6404: This name does not have a type, and must have an explicit type.   [H5T_STD_U16LE]
      call h5dread_f(tb10h_id, H5T_STD_U16LE, tb10h_raw, dims_tb, status)
-------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(412): error #6404: This name does not have a type, and must have an explicit type.   [H5T_IEEE_F64LE]
      call h5dread_f(time_id, H5T_IEEE_F64LE, scantime, dims_time, status)
------------------------------^
compilation aborted for ../USAFSI/USAFSI_amsr2Mod.F90 (code 1)
make: *** [USAFSI_amsr2Mod.o] Error 1

[ERR] Compile failed

I also tried to use mpiifort and mpiicc, but it seems to be the same result. Wondering if anyone could comment on this?
Thanks!

**Comment by bmcandr** [comment, marked as answer]
Hi @tais958,

Thanks for bringing this to our attention. It looks a good chunk of the subroutine `read_amsr2_attributes` in `USAFSI_amsr2Mod.F90` should be within a preprocessor block that checks whether we're compiling with HDF5. This also appears to be an issue in `USAFSI_xcalgmiMod.F90`.

Here are 2 options:

1. Make the following changes:

    * In `USAFSI/USAFSI_amsr2Mod.F90` add `#if (defined USE_HDF5)` to line ~218~ 241 and `#endif` to line 631 (just before `end subroutine read_amsr2_attributes`)
    * In `USAFSI_xcalgmiMod.F90` add `#if (defined USE_HDF5)` to line 236 and `#endif` to line 507 (just before `end subroutine read_xcalgmi_attributes`)

2. Install the optional HDF5 library and compile LDT and LIS with HDF5 enabled.

I will open an issue about this so we can get it fixed.

-----

*Edited to remove initial workaround.*

**Reply by tais958** [reply]
Thanks @bmcandr,

Glad this is helpful for the code development. 
I tried what you suggested, but some errors still occur:

mpiifort -c -O2 -u -traceback -fpe0  -nomixed_str_len_arg -names lowercase -convert big_endian -assume byterecl -DIFC -I/qfs/projects/ops/rh6/esmf/7.1/mod/modO/Linux.intel.64.intelmpi.default -DUSE_INCLUDE_MPI -I/pic/projects/sooty2/tais958/lib/eccodes-2.12.5_intel-16.0.2/include/ -I/share/apps/netcdf/4.3.2/intel/16.0.2/include/  -I./ -I../lib/xmrg/ -I../params/VIC/ -I../USAFSI/ ../USAFSI/USAFSI_amsr2Mod.F90
../USAFSI/USAFSI_amsr2Mod.F90(211): warning #6717: This name has not been given an explicit type.   [FILENAME]
   subroutine read_amsr2_attributes(filename, date10_arr, lat, lon, tb10h, tb10v, &
------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(211): warning #6717: This name has not been given an explicit type.   [DATE10_ARR]
   subroutine read_amsr2_attributes(filename, date10_arr, lat, lon, tb10h, tb10v, &
----------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(211): warning #6717: This name has not been given an explicit type.   [LAT]
   subroutine read_amsr2_attributes(filename, date10_arr, lat, lon, tb10h, tb10v, &
----------------------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(211): warning #6717: This name has not been given an explicit type.   [LON]
   subroutine read_amsr2_attributes(filename, date10_arr, lat, lon, tb10h, tb10v, &
---------------------------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(211): warning #6717: This name has not been given an explicit type.   [TB10H]
   subroutine read_amsr2_attributes(filename, date10_arr, lat, lon, tb10h, tb10v, &
--------------------------------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(211): warning #6717: This name has not been given an explicit type.   [TB10V]
   subroutine read_amsr2_attributes(filename, date10_arr, lat, lon, tb10h, tb10v, &
---------------------------------------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(212): warning #6717: This name has not been given an explicit type.   [TB19H]
                   tb19h, tb19v, tb23h, tb23v, tb37h, tb37v, tb89h, tb89v)
-------------------^
../USAFSI/USAFSI_amsr2Mod.F90(212): warning #6717: This name has not been given an explicit type.   [TB19V]
                   tb19h, tb19v, tb23h, tb23v, tb37h, tb37v, tb89h, tb89v)
--------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(212): warning #6717: This name has not been given an explicit type.   [TB23H]
                   tb19h, tb19v, tb23h, tb23v, tb37h, tb37v, tb89h, tb89v)
---------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(212): warning #6717: This name has not been given an explicit type.   [TB23V]
                   tb19h, tb19v, tb23h, tb23v, tb37h, tb37v, tb89h, tb89v)
----------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(212): warning #6717: This name has not been given an explicit type.   [TB37H]
                   tb19h, tb19v, tb23h, tb23v, tb37h, tb37v, tb89h, tb89v)
-----------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(212): warning #6717: This name has not been given an explicit type.   [TB37V]
                   tb19h, tb19v, tb23h, tb23v, tb37h, tb37v, tb89h, tb89v)
------------------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(212): warning #6717: This name has not been given an explicit type.   [TB89H]
                   tb19h, tb19v, tb23h, tb23v, tb37h, tb37v, tb89h, tb89v)
-------------------------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(212): warning #6717: This name has not been given an explicit type.   [TB89V]
                   tb19h, tb19v, tb23h, tb23v, tb37h, tb37v, tb89h, tb89v)
--------------------------------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(211): error #6404: This name does not have a type, and must have an explicit type.   [FILENAME]
   subroutine read_amsr2_attributes(filename, date10_arr, lat, lon, tb10h, tb10v, &
------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(211): error #6404: This name does not have a type, and must have an explicit type.   [DATE10_ARR]
   subroutine read_amsr2_attributes(filename, date10_arr, lat, lon, tb10h, tb10v, &
----------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(211): error #6404: This name does not have a type, and must have an explicit type.   [LAT]
   subroutine read_amsr2_attributes(filename, date10_arr, lat, lon, tb10h, tb10v, &
----------------------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(211): error #6404: This name does not have a type, and must have an explicit type.   [LON]
   subroutine read_amsr2_attributes(filename, date10_arr, lat, lon, tb10h, tb10v, &
---------------------------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(211): error #6404: This name does not have a type, and must have an explicit type.   [TB10H]
   subroutine read_amsr2_attributes(filename, date10_arr, lat, lon, tb10h, tb10v, &
--------------------------------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(211): error #6404: This name does not have a type, and must have an explicit type.   [TB10V]
   subroutine read_amsr2_attributes(filename, date10_arr, lat, lon, tb10h, tb10v, &
---------------------------------------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(212): error #6404: This name does not have a type, and must have an explicit type.   [TB19H]
                   tb19h, tb19v, tb23h, tb23v, tb37h, tb37v, tb89h, tb89v)
-------------------^
../USAFSI/USAFSI_amsr2Mod.F90(212): error #6404: This name does not have a type, and must have an explicit type.   [TB19V]
                   tb19h, tb19v, tb23h, tb23v, tb37h, tb37v, tb89h, tb89v)
--------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(212): error #6404: This name does not have a type, and must have an explicit type.   [TB23H]
                   tb19h, tb19v, tb23h, tb23v, tb37h, tb37v, tb89h, tb89v)
---------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(212): error #6404: This name does not have a type, and must have an explicit type.   [TB23V]
                   tb19h, tb19v, tb23h, tb23v, tb37h, tb37v, tb89h, tb89v)
----------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(212): error #6404: This name does not have a type, and must have an explicit type.   [TB37H]
                   tb19h, tb19v, tb23h, tb23v, tb37h, tb37v, tb89h, tb89v)
-----------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(212): error #6404: This name does not have a type, and must have an explicit type.   [TB37V]
                   tb19h, tb19v, tb23h, tb23v, tb37h, tb37v, tb89h, tb89v)
------------------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(212): error #6404: This name does not have a type, and must have an explicit type.   [TB89H]
                   tb19h, tb19v, tb23h, tb23v, tb37h, tb37v, tb89h, tb89v)
-------------------------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(212): error #6404: This name does not have a type, and must have an explicit type.   [TB89V]
                   tb19h, tb19v, tb23h, tb23v, tb37h, tb37v, tb89h, tb89v)
--------------------------------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(76): error #6634: The shape matching rules of actual arguments and dummy arguments have been violated.   [DATE0]
         call read_amsr2_attributes(filename, date0, lat0, lon0, tb10h0, tb10v0, &
----------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(76): error #6634: The shape matching rules of actual arguments and dummy arguments have been violated.   [LAT0]
         call read_amsr2_attributes(filename, date0, lat0, lon0, tb10h0, tb10v0, &
-----------------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(76): error #6634: The shape matching rules of actual arguments and dummy arguments have been violated.   [LON0]
         call read_amsr2_attributes(filename, date0, lat0, lon0, tb10h0, tb10v0, &
-----------------------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(76): error #6634: The shape matching rules of actual arguments and dummy arguments have been violated.   [TB10H0]
         call read_amsr2_attributes(filename, date0, lat0, lon0, tb10h0, tb10v0, &
-----------------------------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(76): error #6634: The shape matching rules of actual arguments and dummy arguments have been violated.   [TB10V0]
         call read_amsr2_attributes(filename, date0, lat0, lon0, tb10h0, tb10v0, &
-------------------------------------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(77): error #6634: The shape matching rules of actual arguments and dummy arguments have been violated.   [TB19H0]
                  tb19h0, tb19v0, tb23h0, tb23v0, tb37h0, tb37v0, tb89h0, tb89v0)
------------------^
../USAFSI/USAFSI_amsr2Mod.F90(77): error #6634: The shape matching rules of actual arguments and dummy arguments have been violated.   [TB19V0]
                  tb19h0, tb19v0, tb23h0, tb23v0, tb37h0, tb37v0, tb89h0, tb89v0)
--------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(77): error #6634: The shape matching rules of actual arguments and dummy arguments have been violated.   [TB23H0]
                  tb19h0, tb19v0, tb23h0, tb23v0, tb37h0, tb37v0, tb89h0, tb89v0)
----------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(77): error #6634: The shape matching rules of actual arguments and dummy arguments have been violated.   [TB23V0]
                  tb19h0, tb19v0, tb23h0, tb23v0, tb37h0, tb37v0, tb89h0, tb89v0)
------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(77): error #6634: The shape matching rules of actual arguments and dummy arguments have been violated.   [TB37H0]
                  tb19h0, tb19v0, tb23h0, tb23v0, tb37h0, tb37v0, tb89h0, tb89v0)
--------------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(77): error #6634: The shape matching rules of actual arguments and dummy arguments have been violated.   [TB37V0]
                  tb19h0, tb19v0, tb23h0, tb23v0, tb37h0, tb37v0, tb89h0, tb89v0)
----------------------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(77): error #6634: The shape matching rules of actual arguments and dummy arguments have been violated.   [TB89H0]
                  tb19h0, tb19v0, tb23h0, tb23v0, tb37h0, tb37v0, tb89h0, tb89v0)
------------------------------------------------------------------^
../USAFSI/USAFSI_amsr2Mod.F90(77): error #6634: The shape matching rules of actual arguments and dummy arguments have been violated.   [TB89V0]
                  tb19h0, tb19v0, tb23h0, tb23v0, tb37h0, tb37v0, tb89h0, tb89v0)
--------------------------------------------------------------------------^
compilation aborted for ../USAFSI/USAFSI_amsr2Mod.F90 (code 1)
make: *** [USAFSI_amsr2Mod.o] Error 1

[ERR] Compile failed
-------------------------------------------------------------------

I guess this is due to the variables in the call function for the subroutine `read_amsr2_attributes` were not defined in this consequence. So, I tried to skip the entire subroutine `read_amsr2_attributes` by adding the if statement as you suggested, and it worked. I believe this bug will be fixed in the later version, but it would work as long as I don't this specific data. Is this correct?

Thanks!

**Reply by bmcandr** [reply]
>I guess this is due to the variables in the call function for the subroutine read_amsr2_attributes were not defined in this consequence. So, I tried to skip the entire subroutine read_amsr2_attributes by adding the if statement as you suggested, and it worked.

Sorry, I made a mistake on the line number in my original comment that I failed to update during subsequent edits. See the correction here:

>In USAFSI/USAFSI_amsr2Mod.F90 add #if (defined USE_HDF5) to line ~218~ 241 and #endif to line 631 (just before end subroutine read_amsr2_attributes)

The subroutine argument declarations must not be included in the HDF5 preprocessor block or you will see the type error in your comment. Using the preprocessor check to skip the whole subroutine is an acceptable workaround too.

> I believe this bug will be fixed in the later version, but it would work as long as I don't this specific data. Is this correct?

Correct. LDT should work just fine as long as you're not using the USAFSI data (I suspect not as that is for a specific project of ours). I have already made the fix to one of our support branches (see #1058) and it will eventually be merged into our other branches. Thanks again for your feedback!

**Reply by tais958** [reply]
Great, thanks so much for the prompt reply! I think the three software packages are now successfully compiled on my machine!

**Reply by bmcandr** [reply]
Woohoo! 🎉
